### PR TITLE
[14.0][ADD] stock_warehouse_analytic_purchase

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
Adds to the analytical fields and analytical tags of the lines that are automatically calculated

### CODE REVIEW

File: stock_warehouse_analytic_purchase/models/purchase_order_line.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from odoo import _, models, api
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+
+    @api.depends('order_id.picking_type_id.warehouse_id.account_analytic_id', 'order_id.picking_type_id.warehouse_id.account_analytic_tag_ids')
+    def _compute_account_analytic_id(self):
+        super()._compute_account_analytic_id()
+        for record in self.filtered(lambda x: x.order_id.picking_type_id.warehouse_id.account_analytic_id):
```
Comment (dalonsod) [inline stock_warehouse_analytic_purchase/models/purchase_order_line.py]: According to original method definition (https://github.com/odoo/odoo/blob/f2765d2cab5671a010404c36842bf1b4c4d6350b/addons/purchase/models/purchase.py#L1013) those special purchase lines (section and notes, not regular lines) should be ignored in the process

File: stock_warehouse_analytic_purchase/models/purchase_order_line.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# © 2024 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+
+from odoo import _, models, api
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = 'purchase.order.line'
+
+    @api.depends('order_id.picking_type_id.warehouse_id.account_analytic_id', 'order_id.picking_type_id.warehouse_id.account_analytic_tag_ids')
```
Comment (dalonsod) [inline stock_warehouse_analytic_purchase/models/purchase_order_line.py]: For `tag_ids` there's a separated compute method, please check

File: stock_warehouse_analytic_purchase/models/purchase_order_line.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# © 2024 Solvos Consultoría Informática (<http://www.solvos.es>)
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
+
+from odoo import _, models, api
```
Comment (dalonsod) [inline stock_warehouse_analytic_purchase/models/purchase_order_line.py]: Remove unused import:
```suggestion
from odoo import models, api
```

File: stock_warehouse_analytic_purchase/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Adds to the analytical fields and analytical tags of the lines that are automatically calculated
```
Comment (dalonsod) [inline stock_warehouse_analytic_purchase/readme/DESCRIPTION.rst]: Please rewrite description and make it more detailed (this definition doesn't refer to warehouses, and does not indicate that this new definition has higher priority than default analytic rules)

**Review by dalonsod — CHANGES_REQUESTED** [review]
Not sure if this PR was ready to be reviewed, if not I'm sorry... anyway, I've made some comments to be checked

**Comment by dalonsod** [comment]
OK, I have some performance doubts with the current implementation:

- When `account_analytic_id` and/or `account_analytic_tag_ids` are changed in `stock.warehouse` record, every affected purchase order line (POL) is automatically changed?
- With `stock_warehouse_analytic` already installed and configured for a warehouse, when we install this addon, account and tag fields for POL are updated?

I'm worried about installing (and later configuring) this in a Production environment with a lot of POLs.

Please check both questions, and I'll decide if we move `@api.depends` from current field to simply `order_id.picking_type_id`, which could be less aggressive.

**Comment by ChristianSantamaria** [comment]
`When account_analytic_id and/or account_analytic_tag_ids are changed in stock.warehouse record, every affected purchase order line (POL) is automatically changed?`
Yes, when you modify an account or label, the fields of the lines are modified.

`With stock_warehouse_analytic already installed and configured for a warehouse, when we install this addon, account and tag fields for POL are updated?`
No, when installing the addon they are not recalculated.

**Comment by dalonsod** [comment]
OK, then we'll make the following changes (temporary in a separate commit):

- Move `@api.depends` to only `order_id.picking_type_id`.
- Check that a later change in warehouse doesn't affect existing PO lines.
- Check that functionality still works (for new POLs analytic is proper applied, and if Picking Operation Type for PO is changed, analytic is changed to).

**Comment by ChristianSantamaria** [comment]
`Check that a later change in warehouse doesn't affect existing PO lines.`
Checked not affect

`Check that functionality still works (for new POLs analytic is proper applied, and if Picking Operation Type for PO is changed, analytic is changed to).`
Still work

**Review by dalonsod — APPROVED** [review]
Approved by customer :+1:

**Pull request merged**